### ISSUE TITLE
Different behavior in configuration while monitoring a drive in whodata and realtime

### ISSUE THREAD

**Issue by DaveVG1** [wazuh/wazuh#3934]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 3.9.5 | Syscheck | Agent | - | Windows |

## Description

If a drive is configured to be monitored the following way:
```
<directories check_all="yes" whodata="yes" report_changes="yes">F:\.</directories>
```
Syscheck expects to monitor a folder called `.` inside the `f:\` driver. The Windows event that syscheck receives has a resolved path in the way `f:\folder\file`. That's making that any changes made under the driver to be ignored.

Frequency will still work and it will alert files with the configuration path. For example:
```
** Alert 1567762084.384979: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2019 Sep 06 11:28:04 (Win7) 10.0.0.7->syscheck
Rule: 550 (level 7) -> 'Integrity checksum changed.'
File 'f:\.\fim\fsdfsfdss.txt' checksum changed.
Size changed from '0' to '18'
Old md5sum was: 'd41d8cd98f00b204e9800998ecf8427e'
New md5sum is : '0646848b07060790a7d5579ee6b067ce'
Old sha1sum was: 'da39a3ee5e6b4b0d3255bfef95601890afd80709'
New sha1sum is : 'f7ea3c78f8513bcd6ec1bb51455e13ac2b5fbd85'
Old sha256sum was: 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
New sha256sum is : '4067cd5fd2f96c096fa9b045e441e807de2572f9799c676fdb9a22a163629e88'
Old modification time was: 'Fri Sep  6 11:07:02 2019', now it is 'Fri Sep  6 11:17:43 2019'

Attributes:
 - Size: 18
 - Date: Fri Sep  6 11:17:43 2019
 - User: Administrators (S-1-5-32-544)
 - MD5: 0646848b07060790a7d5579ee6b067ce
 - SHA1: f7ea3c78f8513bcd6ec1bb51455e13ac2b5fbd85
 - SHA256: 4067cd5fd2f96c096fa9b045e441e807de2572f9799c676fdb9a22a163629e88
 - File attributes: ARCHIVE
 - Permissions: 
   Administrators  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   SYSTEM  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Authenticated Users  (ALLOWED) - DELETE, READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Users  (ALLOWED) - READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_READ_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES
```

This is also making that files like `f:\.\$recyclebin\somefile.txt` to provoke an alert. These kind of files should be ignored as stated here:

https://github.com/wazuh/wazuh/blob/6f47eb206f65ecef6053ff77921639161e1a8db6/src/syscheckd/syscheck.h#L153
I think this should be changed to:
`#define check_removed_file(x) ({ strstr(x, "\\$recycle.bin") ? 1 : 0; })`

This configuration works for `realtime` although `report_changes` option won't work (this needs further investigation).

The configuration that works for `whodata` is the following:
```
<directories check_all="yes" whodata="yes" report_changes="yes">F:\\</directories>
```

Nevertheless, this last configuration doesn't work for `realtime`. This is because the path that we store in our internal configuration would be `f:`. When a `realtime` event arrives, syscheck tries to find the folder in the configuration and an additional `\` is appended as stated here:

https://github.com/wazuh/wazuh/blob/6f47eb206f65ecef6053ff77921639161e1a8db6/src/syscheckd/run_realtime.c#L185

The directory won't be found and the event will be ignored.

## Possible solution

I think we should unify the way we store this kind of path in our internal configuration. If the configuration drive name includes one or several `\.` in it, it should be ignored. For instance:

`f:\.\` ---> `f:\\`
`f:\.\.\.\.\.\` ---> `f:\\`

**Comment by antoniomanuelfr** [comment]
I have been testing this behavior and this has been fixed in the 3.12 version of wazuh. 
Using this config `<directories check_all="yes" whodata="yes" report_changes="yes">d:\.</directories>`.
We can see in the logs this message: `syscheck.c:174 at Start_win32_Syscheck(): INFO: (6003): Monitoring directory/file: 'd:\', with options` and if we generate some alerts we can see that the alerts are generated correctly: 
1. Creating a file 
`** Alert 1584951316.1299: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:15:16 (win20161) any->syscheck
Rule: 554 (level 5) -> 'File added to the system.'
File 'd:\helloworld.txt' added
Mode: whodata
`

2. Modifying a file
`** Alert 1584951404.24123: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:16:44 (win20161) any->syscheck
Rule: 550 (level 7) -> 'Integrity checksum changed.'
File 'd:\helloworld.txt' modified
Mode: whodata
Changed attributes: size,mtime,md5,sha1,sha256
Size changed from '0' to '13'
Old modification time was: '1584951314', now it is '1584951404'
Old md5sum was: 'd41d8cd98f00b204e9800998ecf8427e'
New md5sum is : 'd9663cf69535ae1d53e8c066e09d9495'
Old sha1sum was: 'da39a3ee5e6b4b0d3255bfef95601890afd80709'
New sha1sum is : '9bbecb4fa2dde787b7017d41e8b3f718c2a66168'
Old sha256sum was: 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
New sha256sum is : '6478649d227d626b33cd1a31533a6f8fe4789022c45c4360094cd28bd738cba3'
`  
3. Deleting a file: 
`** Alert 1584951506.26000: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:18:26 (win20161) any->syscheck
Rule: 553 (level 7) -> 'File deleted.'
File 'd:\helloworld.txt' deleted
Mode: whodata
`

If we generate the alerts inside a folder: 
1. Creating a file: 
` ** Alert 1584951883.32601: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:24:43 (win20161) any->syscheck
Rule: 554 (level 5) -> 'File added to the system.'
File 'd:\testdir\helloworld.txt' added
Mode: whodata`

2. Modifying file: 
`** Alert 1584952043.35198: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:27:23 (win20161) any->syscheck
Rule: 550 (level 7) -> 'Integrity checksum changed.'
File 'd:\testdir\helloworld.txt' modified
Mode: whodata
Changed attributes: size,mtime,md5,sha1,sha256
Size changed from '0' to '15'
Old modification time was: '1584951882', now it is '1584952044'
Old md5sum was: 'd41d8cd98f00b204e9800998ecf8427e'
New md5sum is : '48b1bb9884e5ad44c2dec1050d7cdcda'
Old sha1sum was: 'da39a3ee5e6b4b0d3255bfef95601890afd80709'
New sha1sum is : '470164836e7534800018f3f1c820eab4b6938b02'
Old sha256sum was: 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
New sha256sum is : 'a25fa4501ff5032aab567fecb9817e74c882091c7a6b04a66e468df84c4b9a5a'
`

3. Deleted file: 
`
** Alert 1584952114.1885: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:28:34 (win20161) any->syscheck
Rule: 553 (level 7) -> 'File deleted.'
File 'd:\testdir\helloworld.txt' deleted
Mode: whodata
`

Now if we monitor the drive like : `<directories check_all="yes" realtime="yes" report_changes="yes">d:\\</directories>` and make the same testing: 

1. Creating file 
`
** Alert 1584952389.8479: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:33:09 (win20161) any->syscheck
Rule: 554 (level 5) -> 'File added to the system.'
File 'd:\testdir\new text document.txt' added
Mode: real-time
`
2. Modifying file
`** Alert 1584952395.3568: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:33:15 (win20161) any->syscheck
Rule: 550 (level 7) -> 'Integrity checksum changed.'
File 'd:\testdir\helloworld.txt' modified
Mode: real-time
Changed attributes: size,mtime,md5,sha1,sha256
Size changed from '0' to '10'
Old modification time was: '1584952391', now it is '1584952397'
Old md5sum was: 'd41d8cd98f00b204e9800998ecf8427e'
New md5sum is : 'fc5e038d38a57032085441e7fe7010b0'
Old sha1sum was: 'da39a3ee5e6b4b0d3255bfef95601890afd80709'
New sha1sum is : '6adfb183a4a2c94a2f92dab5ade762a47889a5a1'
Old sha256sum was: 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
New sha256sum is : '936a185caaa266bb9cbe981e9e05cb78cd732b0b3280eb944412bb6f8f8f07af'
`
3. Delete file 
`
** Alert 1584952491.9548: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:34:51 (win20161) any->syscheck
Rule: 553 (level 7) -> 'File deleted.'
File 'd:\testdir\helloworld.txt' deleted
Mode: real-time
`

If we generate the alerts inside a folder: 

1. Creating file
`** Alert 1584952628.2381: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:37:08 (win20161) any->syscheck
Rule: 554 (level 5) -> 'File added to the system.'
File 'd:\testdir\helloworld.txt' added
Mode: real-time`

2. Modifying file
`** Alert 1584952768.3569: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:39:28 (win20161) any->syscheck
Rule: 550 (level 7) -> 'Integrity checksum changed.'
File 'd:\testdir\helloworld.txt' modified
Mode: real-time
Changed attributes: size,mtime,md5,sha1,sha256
Size changed from '0' to '12'
Old modification time was: '1584952627', now it is '1584952770'
Old md5sum was: 'd41d8cd98f00b204e9800998ecf8427e'
New md5sum is : 'b4e1e8ce33d3bc6186abdee852e20fb0'
Old sha1sum was: 'da39a3ee5e6b4b0d3255bfef95601890afd80709'
New sha1sum is : '423f91670e315b277282933dab93534ea808d378'
Old sha256sum was: 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
New sha256sum is : '74b9b8e04d77731b2e3b60b21ed7344e7743696eaeaa6874ad2f7e9236b78845'
`
3. Deleting file
`** Alert 1584952809.7803: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2020 Mar 23 08:40:09 (win20161) any->syscheck
Rule: 553 (level 7) -> 'File deleted.'
File 'd:\testdir\helloworld.txt' deleted
Mode: real-time
`

Now that this behavior has been fixed, this issue should be closed. 
Regards.

**Comment by antoniomanuelfr** [comment]
The problem that the files that are moved to the recycle bin generate an alert is caused because the whole volume is monitored, even the specific recycle bin of the volume. 
When a file is deleted, two files are generated in the recycle bin, and this files are going to be monitored. This can cause that syscheck open descriptors that are not going to be used.
